diff --git a/crates/edit/src/sys/unix.rs b/crates/edit/src/sys/unix.rs
index 1d51fa6..a354e53 100644
--- a/crates/edit/src/sys/unix.rs
+++ b/crates/edit/src/sys/unix.rs
@@ -439,24 +439,44 @@ pub fn icu_detect_renaming_suffix(arena: &Arena, handle: NonNull<c_void>) -> BSt
             return res;
         }
 
-        // In the versions (63-76) and distributions (Arch/Debian) I tested,
-        // this symbol seems to be always present. This allows us to call `dladdr`.
-        // It's the `UCaseMap::~UCaseMap()` destructor which for some reason isn't
-        // in a namespace. Thank you ICU maintainers for this oversight.
-        let proc = match get_proc_address::<T>(handle, c"_ZN8UCaseMapD1Ev".as_ptr()) {
-            Ok(proc) => proc,
-            Err(_) => return res,
-        };
+        let mut dli_fname: *const c_char = std::ptr::null();
+
+
+        extern "C" fn dl_callback(
+            info: *mut libc::dl_phdr_info,
+            _size: usize,
+            data: *mut c_void,
+        ) -> c_int {
+            let file_path = unsafe { &mut *(data as *mut *const c_char) };
+
+            if info.is_null() || unsafe { (*info).dlpi_name.is_null() } {
+                return 0;
+            }
+
+            let dlpi_str = unsafe { std::ffi::CStr::from_ptr((*info).dlpi_name) };
 
-        // `dladdr` is specific to GNU's libc unfortunately.
-        let mut info: libc::Dl_info = mem::zeroed();
-        let ret = libc::dladdr(proc, &mut info);
-        if ret == 0 {
+            if let Ok(s) = dlpi_str.to_str() {
+                if s.contains("libicuuc") {
+                    *file_path = dlpi_str.as_ptr();
+                    return 1;
+                }
+            }
+
+            0
+        }
+
+        let ret = libc::dl_iterate_phdr(
+                Some(dl_callback),
+                &mut dli_fname as *mut *const c_char as *mut c_void
+            );
+
+        if ret == 0 || dli_fname.is_null() {
             return res;
         }
 
-        // The library path is in `info.dli_fname`.
-        let path = match std::ffi::CStr::from_ptr(info.dli_fname).to_str() {
+
+        // The library path is in `dli_fname`.
+        let path = match std::ffi::CStr::from_ptr(dli_fname).to_str() {
             Ok(name) => name,
             Err(_) => return res,
         };
